File: app/internal/server/routes/terminal_snapshots.go
package routes

import (
	"encoding/json"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/FloSch62/muxus/app/internal/api"
	"github.com/FloSch62/muxus/app/internal/persist"
	"github.com/FloSch62/muxus/app/internal/server"
)

// terminalSnapshotMaxChars mirrors TERMINAL_SNAPSHOT_MAX_CHARS.
const terminalSnapshotMaxChars = 512_000

// snapshotBodyLimit mirrors SNAPSHOT_BODY_LIMIT: serialized scrollback rides
// in a JSON string, where every ESC byte inflates to six characters, so the
// body can be several times the raw snapshot cap.
const snapshotBodyLimit = 4 * 1024 * 1024

type terminalSnapshotEnvelope struct {
	Snapshot *api.TerminalSnapshotRecord `json:"snapshot"`
}

type snapshotSavedResponse struct {
	Saved bool `json:"saved"`
}

// RegisterTerminalSnapshotRoutes mirrors registerTerminalSnapshotRoutes.
func RegisterTerminalSnapshotRoutes(r chi.Router, db *persist.DB) {
	r.Get("/api/terminal-snapshots/{tabId}", func(w http.ResponseWriter, req *http.Request) {
		record, err := db.TerminalSnapshot(chi.URLParam(req, "tabId"))
		if err != nil {
			writeInternalError(w, err)
			return
		}
		server.WriteJSON(w, http.StatusOK, terminalSnapshotEnvelope{Snapshot: record})
	})

	r.Put("/api/terminal-snapshots/{tabId}", func(w http.ResponseWriter, req *http.Request) {
		body, ok := readJSONBody(w, req, snapshotBodyLimit)
		if !ok {
			return
		}
		var probe struct {
			Data *string `json:"data"`
		}
		if err := json.Unmarshal(body, &probe); err != nil || probe.Data == nil ||
			len(*probe.Data) < 1 || len(*probe.Data) > terminalSnapshotMaxChars {
			writeBadRequest(w, "invalid terminal snapshot")
			return
		}
		if err := db.SaveTerminalSnapshot(chi.URLParam(req, "tabId"), *probe.Data); err != nil {
			writeInternalError(w, err)
			return
		}
		server.WriteJSON(w, http.StatusOK, snapshotSavedResponse{Saved: true})
	})
}

File: server/src/routes/terminal-snapshots.ts
import type { FastifyInstance } from 'fastify';
import { z } from 'zod';
import {
  TERMINAL_SNAPSHOT_MAX_CHARS,
  type TerminalSnapshotRecord,
} from '@muxus/shared';
import type { AppContext } from '../app.js';
import { HttpProblem, sendError } from '../util/errors.js';

const snapshotSaveSchema = z.object({
  data: z.string().min(1).max(TERMINAL_SNAPSHOT_MAX_CHARS),
  formatVersion: z.number().int().positive().optional(),
});

// Serialized scrollback rides in a JSON string, where every ESC byte inflates
// to six characters, so the body can be several times the raw snapshot cap.
const SNAPSHOT_BODY_LIMIT = 4 * 1024 * 1024;

export function registerTerminalSnapshotRoutes(app: FastifyInstance, ctx: AppContext): void {
  app.get(
    '/api/terminal-snapshots/:tabId',
    (req): { snapshot: TerminalSnapshotRecord | null } => {
      const { tabId } = req.params as { tabId: string };
      return { snapshot: ctx.database.terminalSnapshot(tabId) ?? null };
    },
  );

  app.put(
    '/api/terminal-snapshots/:tabId',
    { bodyLimit: SNAPSHOT_BODY_LIMIT },
    async (req, reply): Promise<{ saved: boolean } | void> => {
      const { tabId } = req.params as { tabId: string };
      const parsed = snapshotSaveSchema.safeParse(req.body);
      if (!parsed.success) {
        return sendError(reply, new HttpProblem(400, 'invalid terminal snapshot'));
      }
      ctx.database.saveTerminalSnapshot(
        tabId,
        parsed.data.data,
        parsed.data.formatVersion,
      );
      return { saved: true };
    },
  );
}
